torchrl/collectors/collectors.py

```diff
@@ -790,66 +791,38 @@ def iterator(self) -> Iterator[TensorDictBase]:
 
     def _step_and_maybe_reset(self) -> None:
 
-        any_done = False
-        done_map = {}
-        for done_key in self.env.done_keys:
-            done = self._tensordict.get(("next", done_key))
-            truncated = self._tensordict.get(
-                ("next", _replace_last(done_key, "truncated")),
-                None,
-            )
-            done = (done | truncated) if truncated is not None else done
-            any_sub_done = done.any().item()
-            if any_sub_done and self.reset_when_done:
-                # Add this done to the map, we will need it to reset
-                done_map.update({done_key: done.clone()})
-            any_done += any_sub_done
-
         self._tensordict = step_mdp(
             self._tensordict,
             reward_keys=self.env.reward_keys,
             done_keys=self.env.done_keys,
             action_keys=self.env.action_keys,
         )
-
         if not self.reset_when_done:
             return
+        td_reset = self._tensordict.clone(False)
+        any_done = terminated_or_truncated(
+            td_reset,
+            full_done_spec=self.env.output_spec["full_done_spec"],
+            key="_reset",
+        )
 
         if any_done:
             traj_ids = self._tensordict.get(("collector", "traj_ids"))
             traj_ids = traj_ids.clone()
             # collectors do not support passing other tensors than `"_reset"`
             # to `reset()`.
-            td_reset = self._tensordict.select(*done_map.keys())
-            for done_key, done in done_map.items():
-                td_reset.set(_replace_last(done_key, "_reset"), done)
-                del td_reset[done_key]
+            traj_sop = _bring_reset_to_root(td_reset)
```

Comment:
we need to pass the reset keys here because it is very expensive to scan the whole td